internal/dag/builder_test.go

```diff
@@ -1071,6 +1071,124 @@ func TestDAGInsertGatewayAPI(t *testing.T) {
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "basic",
+						Namespace: "projectcontour",
+						Labels: map[string]string{
+							"app":  "contour",
+							"type": "controller",
+						},
+					},
+					Spec: gatewayapi_v1alpha1.HTTPRouteSpec{
+						Hostnames: []gatewayapi_v1alpha1.Hostname{
+							"test.projectcontour.io",
+						},
+						Rules: []gatewayapi_v1alpha1.HTTPRouteRule{{
+							Matches: []gatewayapi_v1alpha1.HTTPRouteMatch{{
+								Path: gatewayapi_v1alpha1.HTTPPathMatch{
+									Type:  "Prefix",
+									Value: "/",
+								},
+							}, {
+								Headers: &gatewayapi_v1alpha1.HTTPHeaderMatch{
+									Type:   "Exact",
+									Values: map[string]string{"foo": "bar"},
+								},
+							}},
+							ForwardTo: []gatewayapi_v1alpha1.HTTPRouteForwardTo{{
+								ServiceName: pointer.StringPtr("kuard"),
+								Port:        gatewayPort(8080),
+							}},
+						}},
+					},
+				},
+			},
+			want: listeners(
+				&Listener{
+					Port: 80,
+					VirtualHosts: virtualhosts(virtualhost("test.projectcontour.io", &Route{
+						PathMatchCondition: prefixString("/"),
+						HeaderMatchConditions: []HeaderMatchCondition{
+							{Name: "foo", Value: "bar", MatchType: "exact", Invert: false},
+						},
+						Clusters: clusters(service(kuardService)),
+					}),
+					),
+				},
+			),
+		},
+		"insert basic single route with single header match, multiple path match": {
+			gateway: gatewayWithSelector,
+			objs: []interface{}{
+				kuardService,
+				&gatewayapi_v1alpha1.HTTPRoute{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "basic",
+						Namespace: "projectcontour",
+						Labels: map[string]string{
+							"app":  "contour",
+							"type": "controller",
+						},
+					},
+					Spec: gatewayapi_v1alpha1.HTTPRouteSpec{
+						Hostnames: []gatewayapi_v1alpha1.Hostname{
+							"test.projectcontour.io",
+						},
+						Rules: []gatewayapi_v1alpha1.HTTPRouteRule{
+							{
+								Matches: []gatewayapi_v1alpha1.HTTPRouteMatch{
+									{
+										Path: gatewayapi_v1alpha1.HTTPPathMatch{
+											Type:  "Prefix",
+											Value: "/blog",
+										},
+									}, {
+										Path: gatewayapi_v1alpha1.HTTPPathMatch{
+											Type:  "Prefix",
+											Value: "/tech",
+										},
+									}, {
+										Headers: &gatewayapi_v1alpha1.HTTPHeaderMatch{
+											Type:   "Exact",
+											Values: map[string]string{"foo": "bar"},
+										},
+									},
+								},
+								ForwardTo: []gatewayapi_v1alpha1.HTTPRouteForwardTo{{
+									ServiceName: pointer.StringPtr("kuard"),
+									Port:        gatewayPort(8080),
+								}},
+							}},
+					},
+				},
+			},
+			want: listeners(
+				&Listener{
+					Port: 80,
+					VirtualHosts: virtualhosts(virtualhost("test.projectcontour.io",
+						&Route{
+							PathMatchCondition: prefixString("/tech"),
+							HeaderMatchConditions: []HeaderMatchCondition{
+								{Name: "foo", Value: "bar", MatchType: "exact", Invert: false},
```

Comment:
Is this correct? From the documentation it seems like there should be three separate routes generated for this test case, one that matches `/tech`, one that matches `/blog` and one that matches the header `foo` with value `bar` combined with the path `/`

Reply:
You are correct. It was wrong. Thank you! Updated.